File: packages/core/tests/timeline.test.ts
import assert from "node:assert/strict";
import { mkdir, mkdtemp, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import path from "node:path";
import { describe, it } from "node:test";

import { listTimelineEntries, readSnapshot } from "../src/store.js";
import { captureTimelineSnapshot } from "../src/timeline.js";
import { buildTimelineUndoPlan } from "../src/timeline-undo.js";
import type { RuntimeOptions } from "../src/runtime-options.js";

async function makeRuntime(): Promise<RuntimeOptions> {
  const root = await mkdtemp(path.join(tmpdir(), "gandalf-timeline-test-"));
  const projectPath = path.join(root, "project");
  const homeDir = path.join(root, "home");
  const storeDir = path.join(root, "store");
  await mkdir(projectPath, { recursive: true });
  await mkdir(homeDir, { recursive: true });
  return { projectPath, homeDir, storeDir };
}

async function writeMcp(projectPath: string, command: string): Promise<void> {
  await writeFile(path.join(projectPath, ".mcp.json"), JSON.stringify({
    mcpServers: {
      github: { transport: "stdio", command }
    }
  }, null, 2));
}

describe("timeline capture", () => {
  it("creates a manual baseline, captures partial MCP+skill changes, and builds MCP-only dry-run undo", async () => {
    const options = await makeRuntime();
    await writeMcp(options.projectPath, "gh-mcp");

    const baseline = await captureTimelineSnapshot(options, {
      captureId: "capture-test",
      snapshotName: "manual-baseline-capture-test"
    });

    assert.equal(baseline.written, true);
    assert.equal(baseline.entry?.eventKind, "baseline");
    assert.equal(baseline.entry?.restoreReadiness, "observe-only");
    assert.equal((await listTimelineEntries(options.storeDir)).length, 1);
    assert.equal((await readSnapshot(options.storeDir, "manual-baseline-capture-test")).manifest.name, "manual-baseline-capture-test");

    await writeMcp(options.projectPath, "gh-mcp-v2");
    const skillDir = path.join(options.homeDir, ".claude", "skills", "react-review");
    await mkdir(skillDir, { recursive: true });
    await writeFile(path.join(skillDir, "SKILL.md"), "# React Review\n");

    const changed = await captureTimelineSnapshot(options, {
      captureId: "capture-test",
      skipUnchanged: true
    });

    assert.equal(changed.written, true);
    assert.equal(changed.entry?.eventKind, "setup_changed");
    assert.equal(changed.entry?.restoreReadiness, "partial");
    assert.equal(changed.entry?.changedSurfaces.some((surface) => surface.kind === "mcp_server" && surface.restorable), true);
    assert.equal(changed.entry?.changedSurfaces.some((surface) => surface.kind !== "mcp_server" && surface.observeOnly), true);

    const undo = await buildTimelineUndoPlan(options.storeDir, changed.entry!.id);
    assert.equal(undo.dryRun, true);
    assert.equal(undo.writesFiles, false);
    assert.equal(undo.writableItems.length, 1);
    assert.equal(undo.writableItems[0].action, "update");
    assert.equal(undo.writableItems[0].serverName, "github");
    assert.ok(undo.observeOnlySurfaces.length >= 1);

    const unchanged = await captureTimelineSnapshot(options, {
      captureId: "capture-test",
      skipUnchanged: true
    });
    assert.equal(unchanged.written, false);
    assert.equal(unchanged.skippedReason, "unchanged");
  });
});

File: crates/gandalf-core/tests/timeline_test.rs
use std::fs;

use gandalf_core::{
    build_timeline_undo_plan, capture_timeline_snapshot, list_timeline_entries, read_snapshot,
    types::RuntimeOptions, BuildTimelineUndoOptions, CaptureTimelineOptions,
    TimelineEntryEventKind, TimelineRestoreReadiness,
};
use tempfile::TempDir;

fn make_runtime() -> (TempDir, RuntimeOptions) {
    let root = TempDir::new().expect("temp dir");
    let project_path = root.path().join("project");
    let home_dir = root.path().join("home");
    let store_dir = root.path().join("store");
    fs::create_dir_all(&project_path).expect("project dir");
    fs::create_dir_all(&home_dir).expect("home dir");
    (
        root,
        RuntimeOptions {
            project_path: project_path.display().to_string(),
            home_dir: home_dir.display().to_string(),
            store_dir: store_dir.display().to_string(),
            agent: None,
            scope: None,
            capture_content: None,
        },
    )
}

fn write_mcp(project_path: &std::path::Path, command: &str) {
    let payload = serde_json::json!({
        "mcpServers": {
            "github": {
                "transport": "stdio",
                "command": command
            }
        }
    });
    fs::write(
        project_path.join(".mcp.json"),
        serde_json::to_string_pretty(&payload).expect("serialize mcp"),
    )
    .expect("write mcp");
}

#[test]
fn creates_baseline_captures_partial_changes_and_builds_mcp_only_dry_run_undo() {
    let (root, options) = make_runtime();
    write_mcp(&root.path().join("project"), "gh-mcp");

    let baseline = capture_timeline_snapshot(
        &options,
        &CaptureTimelineOptions {
            capture_id: Some("capture-test".to_string()),
            snapshot_name: Some("manual-baseline-capture-test".to_string()),
            title: None,
            skip_unchanged: false,
        },
    )
    .expect("baseline capture");

    assert!(baseline.written);
    let entry = baseline.entry.expect("baseline entry");
    assert_eq!(entry.event_kind, TimelineEntryEventKind::Baseline);
    assert_eq!(entry.restore_readiness, TimelineRestoreReadiness::ObserveOnly);
    assert_eq!(
        list_timeline_entries(
            root.path().join("store").as_path(),
            gandalf_core::TimelineListOptions {
                agent: None,
                project_path: None,
                limit: None,
                on_corrupt_entry: None,
            }
        )
        .expect("list timeline")
        .len(),
        1
    );
    assert_eq!(
        read_snapshot(
            root.path().join("store").as_path(),
            "manual-baseline-capture-test",
            None
        )
        .expect("read snapshot")
        .manifest
        .name,
        "manual-baseline-capture-test"
    );

    write_mcp(&root.path().join("project"), "gh-mcp-v2");
    let skill_dir = root
        .path()
        .join("home")
        .join(".claude")
        .join("skills")
        .join("react-review");
    fs::create_dir_all(&skill_dir).expect("skill dir");
    fs::write(skill_dir.join("SKILL.md"), "# React Review\n").expect("skill file");

    let changed = capture_timeline_snapshot(
        &options,
        &CaptureTimelineOptions {
            capture_id: Some("capture-test".to_string()),
            snapshot_name: None,
            title: None,
            skip_unchanged: true,
        },
    )
    .expect("changed capture");

    assert!(changed.written);
    let changed_entry = changed.entry.expect("changed entry");
    assert_eq!(changed_entry.event_kind, TimelineEntryEventKind::SetupChanged);
    assert_eq!(
        changed_entry.restore_readiness,
        TimelineRestoreReadiness::Partial
    );
    assert!(changed_entry.changed_surfaces.iter().any(|surface| {
        surface.kind == "mcp_server" && surface.restorable
    }));
    assert!(changed_entry.changed_surfaces.iter().any(|surface| {
        surface.kind != "mcp_server" && surface.observe_only
    }));

    let undo = build_timeline_undo_plan(
        root.path().join("store").as_path(),
        &changed_entry.id,
        BuildTimelineUndoOptions::default(),
    )
    .expect("undo plan");
    assert!(undo.dry_run);
    assert!(!undo.writes_files);
    assert_eq!(undo.writable_items.len(), 1);
    assert_eq!(undo.writable_items[0].action.as_str(), "update");
    assert_eq!(undo.writable_items[0].server_name, "github");
    assert!(undo.observe_only_surfaces.len() >= 1);

    let unchanged = capture_timeline_snapshot(
        &options,
        &CaptureTimelineOptions {
            capture_id: Some("capture-test".to_string()),
            snapshot_name: None,
            title: None,
            skip_unchanged: true,
        },
    )
    .expect("unchanged capture");
    assert!(!unchanged.written);
    assert_eq!(unchanged.skipped_reason.as_deref(), Some("unchanged"));
}